sound/soc/sof/sof-dai-clk.c

```diff
@@ -0,0 +1,161 @@
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Brent Lu <brent.lu@intel.com>
+//
+// DAI clock layer to support explicit clock control in machine driver.
+//
+
+#include <linux/clk-provider.h>
+#include <linux/clkdev.h>
+#include <sound/sof/dai.h>
+#include "sof-priv.h"
+#include "sof-dai-clk.h"
+
+struct sof_clk {
+	struct clk_hw hw;
+	struct snd_sof_dev *sdev;
+	struct sof_dai_clk *dai_clk;
+};
+
+static int sof_clk_ctrl(struct snd_sof_dev *sdev, struct sof_dai_clk *dai_clk, int enable)
+{
+	struct sof_ipc_dai_clkctrl clkctrl;
+	struct sof_ipc_reply reply;
+	int ret;
+
+	memset(&clkctrl, 0, sizeof(clkctrl));
+
+	/* set IPC DAI clock control */
+	clkctrl.hdr.size = sizeof(clkctrl);
+	clkctrl.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CLKCTRL;
+
+	clkctrl.type = dai_clk->type;
+	clkctrl.dai_index = dai_clk->dai_index;
+
+	switch (dai_clk->type) {
+	case SOF_DAI_INTEL_SSP:
+		clkctrl.ssp.hdr.size = sizeof(clkctrl.ssp);
+		clkctrl.ssp.clk_en = enable ? 1 : 0;
+		clkctrl.ssp.clk_id = dai_clk->clk_id;
+		break;
+	default:
+		dev_err(sdev->dev, "type %d not support\n", dai_clk->type);
+		return -EINVAL;
+	}
+
+	dev_dbg(sdev->dev, "clkctrl: SSP%d clk_en %d clk_id %d\n",
+		clkctrl.dai_index, clkctrl.ssp.clk_en, clkctrl.ssp.clk_id);
+
+	/* send IPC to the DSP */
+	ret = sof_ipc_tx_message(sdev->ipc, clkctrl.hdr.cmd, &clkctrl,
+				 sizeof(clkctrl), &reply, sizeof(reply));
+
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"failed to control DAI clock for type %d index %d\n",
+			dai_clk->type, dai_clk->dai_index);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int sof_clk_prepare(struct clk_hw *hw)
+{
+	struct sof_clk *clkdev = container_of(hw, struct sof_clk, hw);
+	struct snd_sof_dev *sdev = clkdev->sdev;
+
+	return sof_clk_ctrl(sdev, clkdev->dai_clk, 1);
+}
+
+static void sof_clk_unprepare(struct clk_hw *hw)
+{
+	struct sof_clk *clkdev = container_of(hw, struct sof_clk, hw);
+	struct snd_sof_dev *sdev = clkdev->sdev;
+
+	sof_clk_ctrl(sdev, clkdev->dai_clk, 0);
+}
+
+static const struct clk_ops sof_clk_ops = {
+	.prepare = sof_clk_prepare,
+	.unprepare = sof_clk_unprepare,
+};
+
+int snd_sof_dai_clks_probe(struct snd_soc_component *component)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_sof_pdata *pdata = sdev->pdata;
+	struct device *dev = component->dev;
+	int i, ret;
+
+	sdev->num_clks = pdata->num_clks;
+
+	if (sdev->num_clks > SOF_DAI_CLKS)
+		sdev->num_clks = SOF_DAI_CLKS;
+
+	for (i = 0; i < sdev->num_clks; ++i) {
+		struct clk_init_data init = {};
+		struct clk_lookup *dai_clk_lookup;
+		struct sof_clk *clkdev;
+
+		clkdev = devm_kzalloc(dev, sizeof(*clkdev), GFP_KERNEL);
+		if (!clkdev) {
+			ret = -ENOMEM;
+			goto err;
+		}
+
+		init.name = pdata->dai_clks[i].name;
+		init.ops = &sof_clk_ops;
+		init.flags = CLK_GET_RATE_NOCACHE | CLK_SET_RATE_GATE;
```

Comment:
The problem here is I don't know the correct flag for the virtual clock. Need expert to comment the flag.

Reply:
not an expert in the clock API, but how about 0? :-) So far this clock doesn't even change rate, so why cannot it be cached?

Reply:
RATE_GATE might be valid for SCLK they tend to like being stopped in order to be changed, but are we ever going to allow that?
if not then should we use fixed-clock for SOF clocks?